File: dustroute/sim.py
from __future__ import annotations

from collections import deque
from dataclasses import dataclass

from .electrical import (
    DeviceOutputState,
    InstantaneousElectricalState,
    PoweredBlockState,
    component_output_level,
    repeater_input_level,
    repeater_input_pos,
    repeater_output_pos,
    solve_instantaneous,
    torch_support_is_powered,
)
from .model import BlockKind, Pos, World


# Backwards-compatible public name.
BlockPower = PoweredBlockState


@dataclass(frozen=True)
class TickState:
    tick: int
    strengths: dict[Pos, int]
    block_power: dict[Pos, PoweredBlockState]
    repeater_powered: dict[Pos, bool]
    torch_powered: dict[Pos, bool]
    instantaneous_iterations: int = 0

    def strength(self, pos: Pos) -> int:
        return self.strengths.get(pos, 0)

    def power(self, pos: Pos) -> PoweredBlockState:
        return self.block_power.get(pos, PoweredBlockState())

    def weak_power(self, pos: Pos) -> int:
        return self.power(pos).weak

    def strong_power(self, pos: Pos) -> int:
        return self.power(pos).strong

    def block_strength(self, pos: Pos) -> int:
        return self.power(pos).level

    def powered(self, pos: Pos) -> bool:
        return self.strength(pos) > 0 or self.block_strength(pos) > 0


class RedstoneTickSimulator:
    """Redstone model split into instantaneous and delayed phases.

    `settle_instantaneous()` solves dust and stored ordinary-block power while
    repeater/torch outputs are frozen. `advance_tick()` samples component inputs,
    advances those delayed outputs by one abstract tick, then settles again.
    """

    def __init__(self, world: World) -> None:
        self.world = world
        self.tick = 0

        self._repeater_powered: dict[Pos, bool] = {}
        self._queues: dict[Pos, deque[bool]] = {}
        self._torch_powered: dict[Pos, bool] = {}

        for pos, block in world.items():
            if block.kind is BlockKind.REPEATER:
                delay = max(1, min(4, block.delay or 1))
                self._repeater_powered[pos] = False
                self._queues[pos] = deque([False] * delay, maxlen=delay)
            elif block.kind is BlockKind.REDSTONE_TORCH:
                # A newly constructed torch starts lit; its support is sampled
                # during the first delayed component transition.
                self._torch_powered[pos] = True

        self._instantaneous = InstantaneousElectricalState(
            signal_levels={},
            block_power={},
            iterations=0,
            converged=True,
        )
        self._strengths: dict[Pos, int] = {}
        self._block_power: dict[Pos, PoweredBlockState] = {}
        self.settle_instantaneous()

    def _devices(self) -> DeviceOutputState:
        return DeviceOutputState(
            repeater_powered=self._repeater_powered,
            torch_lit=self._torch_powered,
        )

    def settle_instantaneous(self) -> TickState:
        """Recompute zero-delay electrical state without advancing time."""

        self._instantaneous = solve_instantaneous(
            self.world,
            self._devices(),
        )
        # Compatibility fields for previous prototype callers.
        self._strengths = dict(self._instantaneous.signal_levels)
        self._block_power = dict(self._instantaneous.block_power)
        return self.snapshot()

    def _settle(self) -> None:
        """Compatibility alias for the old private method."""
        self.settle_instantaneous()

    def snapshot(self) -> TickState:
        return TickState(
            tick=self.tick,
            strengths=dict(self._instantaneous.signal_levels),
            block_power=dict(self._instantaneous.block_power),
            repeater_powered=dict(self._repeater_powered),
            torch_powered=dict(self._torch_powered),
            instantaneous_iterations=self._instantaneous.iterations,
        )

    def _src(self, pos: Pos) -> int:
        """Compatibility helper for direct component output only."""
        return component_output_level(self.world, pos, self._devices())

    def _rep_in(self, pos: Pos, block) -> Pos | None:
        return repeater_input_pos(self.world, pos)

    def _rep_out(self, pos: Pos, block) -> Pos | None:
        return repeater_output_pos(self.world, pos)

    def _read_repeater(self, pos: Pos | None) -> int:
        if pos is None:
            return 0
        return repeater_input_level(
            self.world,
            pos,
            self._instantaneous,
        )

    def advance_tick(self) -> TickState:
        """Advance delayed component outputs by one abstract redstone tick."""

        next_repeaters: dict[Pos, bool] = {}
        for pos, block in self.world.items():
            if block.kind is not BlockKind.REPEATER:
                continue
            input_pos = repeater_input_pos(self.world, pos)
            requested = (
                input_pos is not None
                and repeater_input_level(
                    self.world,
                    input_pos,
                    self._instantaneous,
                )
                > 0
            )
            queue = self._queues[pos]
            queue.append(bool(requested))
            next_repeaters[pos] = queue[0]

        next_torches: dict[Pos, bool] = {}
        for pos, block in self.world.items():
            if block.kind is not BlockKind.REDSTONE_TORCH:
                continue
            next_torches[pos] = not torch_support_is_powered(
                self.world,
                pos,
                self._instantaneous,
            )

        self._repeater_powered.update(next_repeaters)
        self._torch_powered.update(next_torches)
        self.tick += 1
        return self.settle_instantaneous()

    def step(self) -> TickState:
        """Existing public alias for `advance_tick()`."""
        return self.advance_tick()

File: crates/dustroute-core/src/sim.rs
use std::collections::{BTreeMap, VecDeque};

use crate::connectivity::repeater_input_pos;
use crate::electrical::{
    DeviceOutputState, InstantaneousElectricalState, InstantaneousSolveDidNotConverge,
    PoweredBlockState, repeater_input_level, solve_instantaneous, torch_support_is_powered,
};
use crate::world::{BlockKind, Pos, World};

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct TickState {
    pub tick: u64,
    pub strengths: BTreeMap<Pos, u8>,
    pub block_power: BTreeMap<Pos, PoweredBlockState>,
    pub repeater_powered: BTreeMap<Pos, bool>,
    pub torch_lit: BTreeMap<Pos, bool>,
    pub instantaneous_iterations: usize,
}

impl TickState {
    #[must_use]
    pub fn strength(&self, pos: Pos) -> u8 {
        self.strengths.get(&pos).copied().unwrap_or(0)
    }

    #[must_use]
    pub fn power(&self, pos: Pos) -> PoweredBlockState {
        self.block_power.get(&pos).copied().unwrap_or_default()
    }

    #[must_use]
    pub fn powered(&self, pos: Pos) -> bool {
        self.strength(pos) > 0 || self.power(pos).powered()
    }
}

pub struct RedstoneTickSimulator {
    world: World,
    tick: u64,
    repeater_powered: BTreeMap<Pos, bool>,
    repeater_queues: BTreeMap<Pos, VecDeque<bool>>,
    torch_lit: BTreeMap<Pos, bool>,
    instantaneous: InstantaneousElectricalState,
}

impl RedstoneTickSimulator {
    pub fn new(world: World) -> Result<Self, InstantaneousSolveDidNotConverge> {
        let devices = DeviceOutputState::initially_lit(&world);
        let repeater_queues = world
            .iter()
            .filter(|(_, block)| block.kind == BlockKind::Repeater)
            .map(|(pos, block)| {
                let delay = usize::from(block.delay.unwrap_or(1).clamp(1, 4));
                (*pos, VecDeque::from(vec![false; delay]))
            })
            .collect();
        let instantaneous = solve_instantaneous(&world, &devices, 128)?;
        Ok(Self {
            world,
            tick: 0,
            repeater_powered: devices.repeater_powered,
            repeater_queues,
            torch_lit: devices.torch_lit,
            instantaneous,
        })
    }

    fn devices(&self) -> DeviceOutputState {
        DeviceOutputState {
            repeater_powered: self.repeater_powered.clone(),
            torch_lit: self.torch_lit.clone(),
        }
    }

    pub fn settle_instantaneous(&mut self) -> Result<TickState, InstantaneousSolveDidNotConverge> {
        self.instantaneous = solve_instantaneous(&self.world, &self.devices(), 128)?;
        Ok(self.snapshot())
    }

    #[must_use]
    pub fn snapshot(&self) -> TickState {
        TickState {
            tick: self.tick,
            strengths: self.instantaneous.signal_levels.clone(),
            block_power: self.instantaneous.block_power.clone(),
            repeater_powered: self.repeater_powered.clone(),
            torch_lit: self.torch_lit.clone(),
            instantaneous_iterations: self.instantaneous.iterations,
        }
    }

    pub fn advance_tick(&mut self) -> Result<TickState, InstantaneousSolveDidNotConverge> {
        let mut next_repeaters = BTreeMap::new();
        for (pos, block) in self.world.iter() {
            if block.kind != BlockKind::Repeater {
                continue;
            }
            let requested = repeater_input_pos(&self.world, *pos).is_some_and(|input| {
                repeater_input_level(&self.world, input, &self.instantaneous) > 0
            });
            let queue = self
                .repeater_queues
                .get_mut(pos)
                .expect("repeater queue initialized");
            queue.pop_front();
            queue.push_back(requested);
            next_repeaters.insert(*pos, queue.front().copied().unwrap_or(requested));
        }
        let next_torches = self
            .world
            .iter()
            .filter(|(_, block)| block.kind == BlockKind::RedstoneTorch)
            .map(|(pos, _)| {
                (
                    *pos,
                    !torch_support_is_powered(&self.world, *pos, &self.instantaneous),
                )
            })
            .collect();
        self.repeater_powered = next_repeaters;
        self.torch_lit = next_torches;
        self.tick += 1;
        self.settle_instantaneous()
    }

    pub fn settle_ticks(
        &mut self,
        count: usize,
    ) -> Result<TickState, InstantaneousSolveDidNotConverge> {
        let mut state = self.snapshot();
        for _ in 0..count {
            state = self.advance_tick()?;
        }
        Ok(state)
    }
}

#[cfg(test)]
mod tests {
    use crate::wire::update_wire_shapes;
    use crate::world::{Block, Facing};

    use super::*;

    #[test]
    fn torch_changes_only_on_tick() {
        let mut world = World::new();
        world.set(Pos::new(0, 0, 0), Block::new(BlockKind::Solid));
        let lever = world.place(BlockKind::Lever, Pos::new(-1, 0, 0));
        lever.powered = Some(true);
        lever.support_offset = Some(Pos::new(1, 0, 0));
        let torch = world.place(BlockKind::RedstoneTorch, Pos::new(1, 0, 0));
        torch.facing = Some(Facing::East);
        torch.support_offset = Some(Pos::new(-1, 0, 0));
        let mut simulator = RedstoneTickSimulator::new(world).unwrap();
        assert_eq!(simulator.snapshot().strength(Pos::new(1, 0, 0)), 15);
        simulator.settle_instantaneous().unwrap();
        assert_eq!(simulator.snapshot().strength(Pos::new(1, 0, 0)), 15);
        assert_eq!(
            simulator
                .advance_tick()
                .unwrap()
                .strength(Pos::new(1, 0, 0)),
            0
        );
    }

    #[test]
    fn repeater_refreshes_signal_after_delay() {
        let mut world = World::new();
        world.fill(
            Pos::new(0, 0, 0),
            Pos::new(3, 0, 0),
            Block::new(BlockKind::Solid),
        );
        world.set(Pos::new(0, 1, 0), Block::new(BlockKind::RedstoneBlock));
        world.place(BlockKind::RedstoneWire, Pos::new(1, 1, 0));
        let repeater = world.place(BlockKind::Repeater, Pos::new(2, 1, 0));
        repeater.facing = Some(Facing::East);
        repeater.delay = Some(1);
        world.place(BlockKind::RedstoneWire, Pos::new(3, 1, 0));
        update_wire_shapes(&mut world);
        let mut simulator = RedstoneTickSimulator::new(world).unwrap();
        assert_eq!(simulator.snapshot().strength(Pos::new(3, 1, 0)), 0);
        assert_eq!(
            simulator
                .advance_tick()
                .unwrap()
                .strength(Pos::new(3, 1, 0)),
            15
        );
    }
}
